src/ScopeProvider/scope.ts

```diff
@@ -0,0 +1,254 @@
+    scopedAtom: T,
+    originalAtom: T,
+    implicitScope?: Scope,
+    explicitScope?: Scope,
+  ) {
+    const scopeKey = implicitScope ?? globalScopeKey;
+    if (!inherited.has(scopeKey)) {
+      inherited.set(scopeKey, new WeakMap());
+    }
+    inherited.get(scopeKey)!.set(
+      originalAtom,
+      [
+        scopedAtom, //
+        explicitScope,
+      ].filter(Boolean) as [T, Scope?],
+    );
+  }
+
+  /**
+   * @returns a copy of the atom for derived atoms or the original atom for primitive and writable atoms
+   */
+  function inheritAtom<T>(
+    anAtom: Atom<T>,
+    originalAtom: Atom<T>,
+    implicitScope?: Scope,
+  ) {
+    if (originalAtom.read !== defaultRead) {
+      return cloneAtom(originalAtom, implicitScope);
+    }
+    return anAtom;
+  }
+
+  /**
+   * @returns a scoped copy of the atom
+   */
+  function cloneAtom<T>(originalAtom: Atom<T>, implicitScope?: Scope) {
+    // avoid reading `init` to preserve lazy initialization
+    const scopedAtom: Atom<T> = Object.create(
+      Object.getPrototypeOf(originalAtom),
+      Object.getOwnPropertyDescriptors(originalAtom),
+    );
+
+    if (scopedAtom.read !== defaultRead) {
+      scopedAtom.read = createScopedRead<typeof scopedAtom>(
+        originalAtom.read.bind(originalAtom),
+        implicitScope,
+      );
+    }
+
+    if (
+      isWritableAtom(scopedAtom) &&
+      isWritableAtom(originalAtom) &&
+      scopedAtom.write !== defaultWrite
+    ) {
+      scopedAtom.write = createScopedWrite(
+        originalAtom.write.bind(originalAtom),
+        implicitScope,
+      );
+    }
+
+    // eslint-disable-next-line camelcase
+    scopedAtom.unstable_is = (a: AnyAtom) => {
+      return (
+        a === scopedAtom ||
+        (a.unstable_is?.(originalAtom) ?? a === originalAtom)
+      );
+    };
+
+    return scopedAtom;
+  }
+
+  function createScopedRead<T extends Atom<unknown>>(
+    read: T['read'],
+    implicitScope?: Scope,
+  ): T['read'] {
+    return function scopedRead(get, opts) {
+      return read(
+        function scopedGet(a) {
+          const [scopedAtom] = getAtom(a, implicitScope);
+          return get(scopedAtom);
+        }, //
+        opts,
+      );
+    };
+  }
+
+  function createScopedWrite<T extends AnyWritableAtom>(
+    write: T['write'],
+    implicitScope?: Scope,
+  ): T['write'] {
+    return function scopedWrite(get, set, ...args) {
+      return write(
+        function scopedGet(a) {
+          const [scopedAtom] = getAtom(a, implicitScope);
+          return get(scopedAtom);
+        },
+        function scopedSet(a, ...v) {
+          const [scopedAtom] = getAtom(a, implicitScope);
+          return set(scopedAtom, ...v);
+        },
+        ...args,
+      );
+    };
+  }
+
+  return currentScope;
+}
+
+function isWritableAtom(anAtom: AnyAtom): anAtom is AnyWritableAtom {
+  return 'write' in anAtom;
+}
+
+const { read: defaultRead, write: defaultWrite } = atom<unknown>(null);
```

Comment:
This is a nice hack, but ideally it would be nice if we could avoid this with `unstable_is`.

Reply:
defaultRead and defaultWrite are used to detect when an atom has a custom read or a custom write.

1. This is important when determining whether to copy the atom or use the original. `unstable_is` will handle this case in the future.
2. It is also a performance enhancement since defaultRead's `this` is faster than using getAtom to resolve the atom.

Reply:
Maybe we can avoid checking it later, but in the implementation of this PR, there are some special branches using that comparison, like `inheritAtom`